src/merkle/new_smt/mod.rs

```diff
@@ -0,0 +1,272 @@
+}
+
+impl NewSmt {
+    /// Returns a new [NewSmt].
+    ///
+    /// All leaves in the returned tree are set to [ZERO; 4].
+    pub fn new() -> Self {
+        let root = *EmptySubtreeRoots::entry(NEW_SMT_DEPTH, 0);
+
+        Self {
+            root,
+            leaves: BTreeMap::new(),
+            inner_nodes: BTreeMap::new(),
+        }
+    }
+
+    /// Returns a new [SimpleSmt] instantiated with leaves set as specified by the provided entries.
+    ///
+    /// All leaves omitted from the entries list are set to [ZERO; 4].
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The number of entries exceeds 2^63 entries.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_entries(
+        entries: impl IntoIterator<Item = (NewSmtKey, Word)>,
+    ) -> Result<Self, MerkleError> {
+        // create an empty tree
+        let mut tree = Self::new();
+
+        // compute the max number of entries. We use an upper bound of depth 63 because we consider
+        // passing in a vector of size 2^64 infeasible.
+        let max_num_entries = 2_usize.pow(tree.depth().min(63).into());
+
+        // This being a sparse data structure, the EMPTY_WORD is not assigned to the `BTreeMap`, so
+        // entries with the empty value need additional tracking.
+        let mut key_set_to_zero = BTreeSet::new();
+
+        for (idx, (key, value)) in entries.into_iter().enumerate() {
+            if idx >= max_num_entries {
+                return Err(MerkleError::InvalidNumEntries(max_num_entries));
+            }
+
+            let old_value = tree.update_leaf(key, value);
+
+            if old_value != EMPTY_WORD || key_set_to_zero.contains(&key) {
+                return Err(MerkleError::DuplicateValuesForIndex(
+                    LeafIndex::<NEW_SMT_DEPTH>::from(key).value(),
+                ));
+            }
+
+            if value == EMPTY_WORD {
+                key_set_to_zero.insert(key);
+            };
+        }
+        Ok(tree)
+    }
+}
+
+impl SparseMerkleTree<NEW_SMT_DEPTH> for NewSmt {
+    type Key = NewSmtKey;
+
+    type Value = Word;
+
+    type Leaf = NewSmtLeaf;
+
+    fn root(&self) -> RpoDigest {
+        self.root
+    }
+
+    fn set_root(&mut self, root: RpoDigest) {
+        self.root = root;
+    }
+
+    fn get_inner_node(&self, index: NodeIndex) -> InnerNode {
+        self.inner_nodes.get(&index).cloned().unwrap_or_else(|| {
+            let node = EmptySubtreeRoots::entry(self.depth(), index.depth() + 1);
+
+            InnerNode { left: *node, right: *node }
+        })
+    }
+
+    fn insert_inner_node(&mut self, index: NodeIndex, inner_node: InnerNode) {
+        self.inner_nodes.insert(index, inner_node);
+    }
+
+    fn insert_value(&mut self, key: Self::Key, value: Self::Value) -> Option<Self::Value> {
+        let leaf_index: LeafIndex<NEW_SMT_DEPTH> = key.into();
+        match self.leaves.get_mut(&leaf_index.value()) {
```

Comment:
There is a special case we need to handle here: `value == EMPTY_VALUE`. In this case, we need to return the tree into the state it was as if the key wasn't inserted in the first place. This implies:

1. For a single leaf - remove the leaf.
2. For multiple leaf:
  a. If there are more than 2 entries in the leaf - remove the entry associated with the key.
  b. If there are exactly 2 entries, remove the relevant entry and convert the leaf into a single leaf.

I would probably create a couple of helper functions to handle these cases separately (would help with readability).

Reply:
I handled the case, but was able to make `perform_remove()` clean. For `insert()`, I moved it to `SmtLeaf`; but you can't do that for `remove()`, since in one case, "removing the leaf" involves removing the element from `Smt.leaves`.

I'll write the tests, and then find a better way to clean it up

Reply:
How about something like this:
```Rust
fn perform_remove(&mut self, key: SmtKey) -> Option<Word> {
    let leaf_index: LeafIndex<SMT_DEPTH> = key.into();

    if let Some(leaf) = self.leaves.get_mut(&leaf_index.value()) {
        let (old_value, is_empty) leaf.remove(key);
        if is_empty {
            self.leaves.remove(&leaf_index.value());
        }
        old_value
    }  else {
        // there's nothing stored at the leaf; nothing to update
        None
    }
}
```

Reply:
done!